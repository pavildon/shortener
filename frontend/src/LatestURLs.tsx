import { FC, useEffect, useState } from "react"
import {URLList} from "./types"

type ListParams = {
  perPage: number,
  error: string | null,
  urls: URLList | null,
  getURLsFn: (skip: number) => void
  baseURL: string
}

export const LatestURLs: FC<ListParams> = ({ perPage, error, urls, getURLsFn, baseURL }) => {

  const [skip, setSkip] = useState<number>(0)

  useEffect(() => {
    getURLsFn(skip)
  }, [skip])

  return (
    <div className="w3-container w3-panel">
      <h4>latests URLs</h4>
      {!error ? (<>
        {urls ? (<>
          <div className="w3-left-align w3-monospace">
            <ol start={skip + 1}
              style={{
                minHeight: "315px",
                listStyleType: "decimal-leading-zero",
                padding: "0px"
              }}>
              {urls.items.map((u, i) => (
                <li key={i}
                  style={{
                    whiteSpace: "nowrap",
                    borderBottom: "1px solid #ddd",
                    padding: "4px",
                    listStylePosition: "inside",
                    textOverflow: "ellipsis",
                    overflow: "hidden"
                  }}>
                  <a href={u.url}>{baseURL}/{u._id}</a>
                  <i className="w3-light-gray w3-margin">{u.url}</i>
                </li>
              ))}
            </ol>
          </div>
          <button
            data-testid="prev"
            className="w3-btn"
            style={{ visibility: (skip <= 0 ? "hidden" : "visible") }}
            onClick={() => setSkip(skip - perPage)}>← prev</button>
          <button
            data-testid="next"
            className="w3-btn"
            style={{ visibility: (((skip + perPage) > urls.total) ? "hidden" : "visible") }}
            onClick={() => setSkip(skip + perPage)}>next →</button>
        </>) : (<i>loading... </i>)
        }
      </>) : (<div className="w3-panel w3-pale-red">{error}</div>)}
    </div>
  )
}
