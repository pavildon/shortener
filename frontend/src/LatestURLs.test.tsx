import { LatestURLs } from "./LatestURLs"
import renderer from "react-test-renderer"
import { act, fireEvent, render, waitFor, screen } from "@testing-library/react"

it("should show loading...", async () => {
  const get = jest.fn(
    async () => { })

  const c = renderer.create(<LatestURLs
    perPage={10}
    urls={null}
    error={null}
    getURLsFn={get}
    baseURL="https://pbid.io" />)

  var tree = c.toJSON()
  expect(tree).toMatchSnapshot()
})

it("should show 1 url", async () => {
  const get = jest.fn(
    async () => { })

  const c = renderer.create(<LatestURLs
    perPage={10}
    urls={{ total: 1, items: [{ _id: "abcdefgh", url: "www.google.com" }] }}
    error={null}
    getURLsFn={get}
    baseURL="https://pbid.io" />)

  var tree = c.toJSON()
  expect(tree).toMatchSnapshot()
})

it("should show 10 url and the next button", async () => {

  const c = renderer.create(<LatestURLs
    perPage={10}
    urls={{ total: url100.length, items: url100.slice(0, 10) }}
    error={null}
    getURLsFn={() => { }}
    baseURL="https://pbid.io" />)

  var tree = c.toJSON()
  expect(tree).toMatchSnapshot()
})

it("should call get to show next and previous page", async () => {
  const get = jest.fn(
    async () => { }
  )

  await act(async () => {
    render(<LatestURLs
      perPage={10}
      urls={{ total: url100.length, items: url100.slice(0, 10) }}
      error={null}
      getURLsFn={get}
      baseURL="https://pbid.io" />)
  })

  await (act(async () => {
    fireEvent.click(await waitFor(() => screen.getByTestId("next")))
  }))

  expect(get).toBeCalledTimes(2)
  expect(get).nthCalledWith(1, 0)
  expect(get).nthCalledWith(2, 10)

  await (act(async () => {
    fireEvent.click(await waitFor(() => screen.getByTestId("prev")))
  }))

  expect(get).toBeCalledTimes(3)
  expect(get).nthCalledWith(3, 0)
})

const pad = (num: number, size: number) => {
  var nums = num.toString();
  while (nums.length < size) {
    nums = "0" + nums
  }
  return nums;
}

const genURLs = (n: number) => {
  const urls = []

  for (var i = 0; i < n; i++) {
    urls.push({ _id: pad(i, 8), url: `https://${pad(i, 12)}.com` })
  }

  return urls
}

const url100 = genURLs(100)


