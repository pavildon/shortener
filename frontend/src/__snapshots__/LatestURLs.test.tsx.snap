// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should show 1 url 1`] = `
<div
  className="w3-container w3-panel"
>
  <h4>
    latests URLs
  </h4>
  <div
    className="w3-left-align w3-monospace"
  >
    <ol
      start={1}
      style={
        Object {
          "listStyleType": "decimal-leading-zero",
          "minHeight": "315px",
          "padding": "0px",
        }
      }
    >
      <li
        style={
          Object {
            "borderBottom": "1px solid #ddd",
            "listStylePosition": "inside",
            "overflow": "hidden",
            "padding": "4px",
            "textOverflow": "ellipsis",
            "whiteSpace": "nowrap",
          }
        }
      >
        <a
          href="www.google.com"
        >
          https://pbid.io
          /
          abcdefgh
        </a>
        <i
          className="w3-light-gray w3-margin"
        >
          www.google.com
        </i>
      </li>
    </ol>
  </div>
  <button
    className="w3-btn"
    data-testid="prev"
    onClick={[Function]}
    style={
      Object {
        "visibility": "hidden",
      }
    }
  >
    ← prev
  </button>
  <button
    className="w3-btn"
    data-testid="next"
    onClick={[Function]}
    style={
      Object {
        "visibility": "hidden",
      }
    }
  >
    next →
  </button>
</div>
`;

exports[`should show 10 url and the next button 1`] = `
<div
  className="w3-container w3-panel"
>
  <h4>
    latests URLs
  </h4>
  <div
    className="w3-left-align w3-monospace"
  >
    <ol
      start={1}
      style={
        Object {
          "listStyleType": "decimal-leading-zero",
          "minHeight": "315px",
          "padding": "0px",
        }
      }
    >
      <li
        style={
          Object {
            "borderBottom": "1px solid #ddd",
            "listStylePosition": "inside",
            "overflow": "hidden",
            "padding": "4px",
            "textOverflow": "ellipsis",
            "whiteSpace": "nowrap",
          }
        }
      >
        <a
          href="https://000000000000.com"
        >
          https://pbid.io
          /
          00000000
        </a>
        <i
          className="w3-light-gray w3-margin"
        >
          https://000000000000.com
        </i>
      </li>
      <li
        style={
          Object {
            "borderBottom": "1px solid #ddd",
            "listStylePosition": "inside",
            "overflow": "hidden",
            "padding": "4px",
            "textOverflow": "ellipsis",
            "whiteSpace": "nowrap",
          }
        }
      >
        <a
          href="https://000000000001.com"
        >
          https://pbid.io
          /
          00000001
        </a>
        <i
          className="w3-light-gray w3-margin"
        >
          https://000000000001.com
        </i>
      </li>
      <li
        style={
          Object {
            "borderBottom": "1px solid #ddd",
            "listStylePosition": "inside",
            "overflow": "hidden",
            "padding": "4px",
            "textOverflow": "ellipsis",
            "whiteSpace": "nowrap",
          }
        }
      >
        <a
          href="https://000000000002.com"
        >
          https://pbid.io
          /
          00000002
        </a>
        <i
          className="w3-light-gray w3-margin"
        >
          https://000000000002.com
        </i>
      </li>
      <li
        style={
          Object {
            "borderBottom": "1px solid #ddd",
            "listStylePosition": "inside",
            "overflow": "hidden",
            "padding": "4px",
            "textOverflow": "ellipsis",
            "whiteSpace": "nowrap",
          }
        }
      >
        <a
          href="https://000000000003.com"
        >
          https://pbid.io
          /
          00000003
        </a>
        <i
          className="w3-light-gray w3-margin"
        >
          https://000000000003.com
        </i>
      </li>
      <li
        style={
          Object {
            "borderBottom": "1px solid #ddd",
            "listStylePosition": "inside",
            "overflow": "hidden",
            "padding": "4px",
            "textOverflow": "ellipsis",
            "whiteSpace": "nowrap",
          }
        }
      >
        <a
          href="https://000000000004.com"
        >
          https://pbid.io
          /
          00000004
        </a>
        <i
          className="w3-light-gray w3-margin"
        >
          https://000000000004.com
        </i>
      </li>
      <li
        style={
          Object {
            "borderBottom": "1px solid #ddd",
            "listStylePosition": "inside",
            "overflow": "hidden",
            "padding": "4px",
            "textOverflow": "ellipsis",
            "whiteSpace": "nowrap",
          }
        }
      >
        <a
          href="https://000000000005.com"
        >
          https://pbid.io
          /
          00000005
        </a>
        <i
          className="w3-light-gray w3-margin"
        >
          https://000000000005.com
        </i>
      </li>
      <li
        style={
          Object {
            "borderBottom": "1px solid #ddd",
            "listStylePosition": "inside",
            "overflow": "hidden",
            "padding": "4px",
            "textOverflow": "ellipsis",
            "whiteSpace": "nowrap",
          }
        }
      >
        <a
          href="https://000000000006.com"
        >
          https://pbid.io
          /
          00000006
        </a>
        <i
          className="w3-light-gray w3-margin"
        >
          https://000000000006.com
        </i>
      </li>
      <li
        style={
          Object {
            "borderBottom": "1px solid #ddd",
            "listStylePosition": "inside",
            "overflow": "hidden",
            "padding": "4px",
            "textOverflow": "ellipsis",
            "whiteSpace": "nowrap",
          }
        }
      >
        <a
          href="https://000000000007.com"
        >
          https://pbid.io
          /
          00000007
        </a>
        <i
          className="w3-light-gray w3-margin"
        >
          https://000000000007.com
        </i>
      </li>
      <li
        style={
          Object {
            "borderBottom": "1px solid #ddd",
            "listStylePosition": "inside",
            "overflow": "hidden",
            "padding": "4px",
            "textOverflow": "ellipsis",
            "whiteSpace": "nowrap",
          }
        }
      >
        <a
          href="https://000000000008.com"
        >
          https://pbid.io
          /
          00000008
        </a>
        <i
          className="w3-light-gray w3-margin"
        >
          https://000000000008.com
        </i>
      </li>
      <li
        style={
          Object {
            "borderBottom": "1px solid #ddd",
            "listStylePosition": "inside",
            "overflow": "hidden",
            "padding": "4px",
            "textOverflow": "ellipsis",
            "whiteSpace": "nowrap",
          }
        }
      >
        <a
          href="https://000000000009.com"
        >
          https://pbid.io
          /
          00000009
        </a>
        <i
          className="w3-light-gray w3-margin"
        >
          https://000000000009.com
        </i>
      </li>
    </ol>
  </div>
  <button
    className="w3-btn"
    data-testid="prev"
    onClick={[Function]}
    style={
      Object {
        "visibility": "hidden",
      }
    }
  >
    ← prev
  </button>
  <button
    className="w3-btn"
    data-testid="next"
    onClick={[Function]}
    style={
      Object {
        "visibility": "visible",
      }
    }
  >
    next →
  </button>
</div>
`;

exports[`should show loading... 1`] = `
<div
  className="w3-container w3-panel"
>
  <h4>
    latests URLs
  </h4>
  <i>
    loading... 
  </i>
</div>
`;
